### ISSUE TITLE
EMA update on CosineCodebook

### ISSUE THREAD

**Issue by le4m** [lucidrains/vector-quantize-pytorch#26]
The original VIT-VQGAN paper does not seem to use EMA update for codebook learning since their codebook is unit-normalized vectors.

Particularly, to my understanding, EMA update does not quite make sense. 

What's your take on this? Should we NOT use EMA update with CosineCodebook?

**Comment by pengzhangzhi** [comment]
Would you like to explain why esm does not work for unit-normalized codebook?

**Comment by Saltychtao** [comment]
I found when using EMA for cosine code book, the l2-norm of the input to the vq module would grow gradually, from 22 -> 20000, leading to growing training loss. Has anyone met this problem?

**Comment by lucidrains** [comment]
@le4m yes i do think you may be correct

i just reread the [vit-vqgan](https://arxiv.org/abs/2110.04627) paper and they are only using a commitment loss with a weight of `0.25`

**Comment by Saltychtao** [comment]
> I found when using EMA for cosine code book, the l2-norm of the input to the vq module would grow gradually, from 22 -> 20000, leading to growing training loss. Has anyone met this problem?

In case anyone else has this problem, I add a `layernorm` layer after the `vq_in` projection, and the growing norm problem is largely solved.

**Comment by jzhang38** [comment]
@Saltychtao  I also encounter a similar issue. Does vq_in refer to VectorQuantize.project_in?

**Comment by Saltychtao** [comment]
> @Saltychtao I also encounter a similar issue. Does vq_in refer to VectorQuantize.project_in?

Yes.

**Comment by santisy** [comment]
> > I found when using EMA for cosine code book, the l2-norm of the input to the vq module would grow gradually, from 22 -> 20000, leading to growing training loss. Has anyone met this problem?
> 
> In case anyone else has this problem, I add a `layernorm` layer after the `vq_in` projection, and the growing norm problem is largely solved.

Hi, just want to make sure that the current vesion of implementation here seems to put one normalization (l2norm) after the project_in. I also encounter the training loss explosion issue somehow at current version

**Comment by lucidrains** [comment]
@santisy want to try turning [this](https://github.com/lucidrains/vector-quantize-pytorch/commit/b38a4c2d1c39bc1b1dae793e3f216cccd10613b6#diff-8bcd9c958b614ce130dc4091c094d4cfcc2023716823c181e3a9edaddbcd433dR714) on

let me know if it helps

**Comment by LindgeW** [comment]
> > I found when using EMA for cosine code book, the l2-norm of the input to the vq module would grow gradually, from 22 -> 20000, leading to growing training loss. Has anyone met this problem?
> 
> In case anyone else has this problem, I add a `layernorm` layer after the `vq_in` projection, and the growing norm problem is largely solved.

@Saltychtao Hi, if a `layernorm` layer is added after project_in, is the l2norm operation necessary for l2-normed  Euclidean codebooks (e.g., ||l2(z) - l2(e)||^2_2)?
